File: backend/custom_addons/document_flow_ext/controllers/document_flow_api.py
# -*- coding: utf-8 -*-
from odoo import http, fields
from odoo.http import request
from odoo.exceptions import UserError

class DocumentFlowApi(http.Controller):
    """
    API cho module document_flow_ext.
    Cho phép React frontend thao tác với văn bản và luồng xử lý.
    """

    # -------------------------------------------------------------------------
    # 1️⃣ LẤY DANH SÁCH ROUTE CỦA MỘT VĂN BẢN
    # -------------------------------------------------------------------------
    @http.route('/web/api/document/<int:doc_id>/routes', type='json', auth='user', methods=['GET'])
    def get_document_routes(self, doc_id):
        """
        Trả về danh sách các bước xử lý (route) của văn bản.
        """
        document = request.env['dms.file'].sudo().browse(doc_id)
        if not document.exists():
            return {'error': 'Không tìm thấy văn bản'}

        routes = request.env['document.route'].sudo().search_read(
            [('document_id', '=', doc_id)],
            [
                'id', 'sequence', 'from_department_id', 'to_department_id',
                'assigned_to', 'action_type', 'state',
                'date_sent', 'date_received', 'processed_date', 'note'
            ],
            order='sequence asc'
        )

        # Format cho React
        for r in routes:
            if r.get('from_department_id'):
                r['from_department_name'] = r['from_department_id'][1]
                r['from_department_id'] = r['from_department_id'][0]
            if r.get('to_department_id'):
                r['to_department_name'] = r['to_department_id'][1]
                r['to_department_id'] = r['to_department_id'][0]
            if r.get('assigned_to'):
                r['assigned_to_name'] = r['assigned_to'][1]
                r['assigned_to'] = r['assigned_to'][0]

        return {'document_id': doc_id, 'routes': routes}

    # -------------------------------------------------------------------------
    # 2️⃣ GỬI LỆNH “CHUYỂN BƯỚC” (ADVANCE)
    # -------------------------------------------------------------------------
    @http.route('/web/api/document/<int:doc_id>/advance', type='json', auth='user', methods=['POST'])
    def advance_document_route(self, doc_id):
        """
        Chuyển văn bản sang bước tiếp theo trong luồng xử lý.
        """
        try:
            request.env['document.route'].sudo().advance_route(doc_id)
            return {'success': True, 'message': 'Đã chuyển sang bước tiếp theo.'}
        except UserError as e:
            return {'success': False, 'error': str(e)}
        except Exception as e:
            return {'success': False, 'error': f'Lỗi hệ thống: {e}'}

    # -------------------------------------------------------------------------
    # 3️⃣ TẠO MỚI ROUTE CHO VĂN BẢN
    # -------------------------------------------------------------------------
    @http.route('/web/api/document/<int:doc_id>/routes/create', type='json', auth='user', methods=['POST'])
    def create_document_route(self, doc_id, **payload):
        """
        Tạo mới một bước luồng xử lý cho văn bản.
        Dữ liệu gửi từ React dạng JSON:
        {
          "sequence": 10,
          "from_department_id": 1,
          "to_department_id": 2,
          "assigned_to": 5,
          "action_type": "approve",
          "note": "Trình lên phòng giám đốc"
        }
        """
        document = request.env['dms.file'].sudo().browse(doc_id)
        if not document.exists():
            return {'success': False, 'error': 'Không tìm thấy văn bản'}

        try:
            new_route = request.env['document.route'].sudo().create({
                'document_id': doc_id,
                'sequence': payload.get('sequence', 10),
                'from_department_id': payload.get('from_department_id'),
                'to_department_id': payload.get('to_department_id'),
                'assigned_to': payload.get('assigned_to'),
                'action_type': payload.get('action_type', 'review'),
                'note': payload.get('note', ''),
                'state': 'pending',
                'date_sent': fields.Datetime.now(),
            })
            return {
                'success': True,
                'message': 'Đã thêm bước xử lý mới.',
                'route_id': new_route.id
            }
        except Exception as e:
            return {'success': False, 'error': f'Lỗi khi tạo route: {e}'}

File: frontend/hrms_frontend/src/services/api/documentFlowAPI.js
//frontend\hrms_frontend\src\services\api\documentFlowAPI.js
import axiosInstance from "../../utils/axios_instance";
import URL from "../../utils/url";

/**
 * Lấy danh sách các bước xử lý (routes) của một văn bản
 * @param {number} documentId - ID của văn bản (model: dms.file)
 * @returns {Promise<Array>} - Danh sách route
 */
export const getDocumentRoutes = async (documentId) => {
  try {
    const response = await axiosInstance.post(
      `/api/document/${documentId}/routes`,
      {}
    );
    if (response.data.error) throw new Error(response.data.error);
    return response.data.result || response.data.routes || [];
  } catch (error) {
    console.error("❌ Lỗi khi lấy danh sách route:", error);
    throw new Error(error.message || "Không thể lấy danh sách route.");
  }
};

/**
 * Gửi lệnh chuyển bước xử lý (advance)
 * @param {number} documentId - ID của văn bản
 * @returns {Promise<object>} - Kết quả trả về từ Odoo
 */
export const advanceDocumentRoute = async (documentId) => {
  try {
    const response = await axiosInstance.post(
      `/api/document/${documentId}/advance`,
      {}
    );
    if (response.data.error) throw new Error(response.data.error);
    return response.data.result || response.data;
  } catch (error) {
    console.error("❌ Lỗi khi chuyển bước văn bản:", error);
    throw new Error(error.message || "Không thể chuyển bước văn bản.");
  }
};

/**
 * Tạo mới một bước xử lý (route) cho văn bản
 * @param {number} documentId - ID của văn bản
 * @param {object} routeData - Dữ liệu route
 * {
 *   sequence: 10,
 *   from_department_id: 1,
 *   to_department_id: 2,
 *   assigned_to: 5,
 *   action_type: "approve",
 *   note: "Trình lên phòng giám đốc"
 * }
 */
export const createDocumentRoute = async (documentId, routeData) => {
  try {
    const response = await axiosInstance.post(
      `/api/document/${documentId}/routes/create`,
      routeData
    );
    if (response.data.error) throw new Error(response.data.error);
    return response.data.result || response.data;
  } catch (error) {
    console.error("❌ Lỗi khi tạo route mới:", error);
    throw new Error(error.message || "Không thể tạo route mới.");
  }
};
